### PR TITLE
[BEAM-6623] add DataflowValidatesRunner batch tests for python 3.6 an…

### PR DESCRIPTION
**Pull request by Juta**
…d 3.7

This pr adds it suites for python 3.6. This is is part of a series of PRs with goal to make Apache Beam PY3 compatible. The proposal with the outlined approach has been documented here: https://s.apache.org/beam-python-3.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by Juta** [comment]
Run Python Dataflow ValidatesRunner

**Comment by Juta** [comment]
Run Python Dataflow ValidatesRunner

**Comment by Juta** [comment]
@tvalentyn I tried adding the validatesrunner tests but they don't appear in the logs. Maybe there is a syntax error when referring to the tests  as such: `tasks(':beam-sdks-python:test-suites:dataflow:py35:validatesRunnerBatchTests')`?

File: sdks/python/test-suites/dataflow/build.gradle [hunk 1]
```diff
@@ -45,3 +45,23 @@ task preCommitIT(dependsOn: ['sdist', 'installGcpTest']) {
     }
   }
 }
+
```
Comment (tvalentyn) [inline sdks/python/test-suites/dataflow/build.gradle]: Did you mean to modify `dataflow/py37/build.gradle` instead of this file?

**Comment by tvalentyn** [comment]
I think you also  need to modify `settings.gradle`, see https://github.com/apache/beam/pull/8564/files.

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
@tvalentyn PTAL

**Comment by fredo838** [comment]
Run Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunnerChangeTest

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
When changing [.test-infra/jenkins/job_PostCommit_Python_ValidatesRunner_Dataflow.groovy](https://github.com/apache/beam/blob/af3f94da0950da3372c692c78dd81e0a8fe389aa/.test-infra/jenkins/job_PostCommit_Python_ValidatesRunner_Dataflow.groovy#L32-L36) to not include ValidateRunnerBatchTests and ~StreamingTests, they are still executed on Jenkins when running "Run Python Dataflow ValidatesRunner". What am I missing? @tvalentyn

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Python PostCommit

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Python Postcommit

**Comment by fredo838** [comment]
Run Python PostCommit

**Comment by fredo838** [comment]
Run seed job

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

File: .test-infra/jenkins/job_PostCommit_Python_ValidatesRunner_Dataflow.groovy [hunk 2]
```diff
@@ -32,8 +32,7 @@ PostcommitJobBuilder.postCommitJob('beam_PostCommit_Py_VR_Dataflow', 'Run Python
   steps {
     gradle {
       rootBuildScriptDir(commonJobProperties.checkoutDir)
-      tasks(':beam-sdks-python:validatesRunnerBatchTests')
-      tasks(':beam-sdks-python:validatesRunnerStreamingTests')
+      tasks(':python3ValidatesRunner')
```
Comment (tvalentyn) [inline .test-infra/jenkins/job_PostCommit_Python_ValidatesRunner_Dataflow.groovy]: Looks like this removes Py2 validates-runner suites? We should keep them.

File: project-mappings [hunk 3]
```diff
@@ -2,13 +2,15 @@
 :beam-vendor-sdks-java-extensions-protobuf :vendor:sdks-java-extensions-protobuf
 :beam-vendor-guava-20_0 :vendor:guava-20_0
 :beam-vendor-grpc-1_13_1 :vendor:grpc-1_13_1
-:beam-sdks-python-test-suites-tox-py36 :sdks:python:test-suites:tox:py37
+:beam-sdks-python-test-suites-tox-py37 :sdks:python:test-suites:tox:py37
```
Comment (tvalentyn) [inline project-mappings]: Let's do consistent ordering of test suites (35, 36, 37)

File: sdks/python/test-suites/dataflow/py37/build.gradle [hunk 4]
```diff
@@ -69,3 +69,23 @@ task validatesRunnerBatchTests(dependsOn: ['installGcpTest', 'sdist']) {
...
+                  "streaming": "true",
+                  "sdk_location": "${project.buildDir}/apache-beam.tar.gz",
+                  "worker_jar": dataflowWorkerJar]
+    def cmdArgs = project.mapToArgString(argMap)
+    exec {
+      executable 'sh'
+      args '-c', ". ${project.ext.envdir}/bin/activate && ${runScriptsDir}/run_integration_test.sh $cmdArgs"
+    }
+  }
+}
```
Comment (tvalentyn) [inline sdks/python/test-suites/dataflow/py37/build.gradle]: Let's add the end-of-line as in other files.

File: project-mappings [hunk 5]
```diff
@@ -2,14 +2,16 @@
 :beam-vendor-sdks-java-extensions-protobuf :vendor:sdks-java-extensions-protobuf
 :beam-vendor-guava-20_0 :vendor:guava-20_0
 :beam-vendor-grpc-1_13_1 :vendor:grpc-1_13_1
-:beam-sdks-python-test-suites-tox-py36 :sdks:python:test-suites:tox:py37
-:beam-sdks-python-test-suites-tox-py36 :sdks:python:test-suites:tox:py36
-:beam-sdks-python-test-suites-tox-py35 :sdks:python:test-suites:tox:py35
 :beam-sdks-python-test-suites-direct-py35 :sdks:python:test-suites:direct:py35
 :beam-sdks-python-test-suites-direct-py36 :sdks:python:test-suites:direct:py36
+:beam-sdks-python-test-suites-direct-py37 :sdks:python:test-suites:direct:py37
```
Comment (fredo838) [inline project-mappings]: Do we even need to to add this here?

File: .test-infra/jenkins/job_PostCommit_Python_ValidatesRunner_Dataflow.groovy (hunk 2 above)
Comment (fredo838) [reply .test-infra/jenkins/job_PostCommit_Python_ValidatesRunner_Dataflow.groovy]: I readded them  (I only removed them to test Jenkins behaviour, turns out `Run Seed Job` is what I needed).

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
@tvalentyn I tried to refactor the code of ValidatesRunnerBatchTest to be defined in BeamModulePlugin.groovy (only for 3.5 for now), but I can't get the dependencies to work. I tried to do this based on [this](https://issues.apache.org/jira/browse/BEAM-6623?focusedCommentId=16847059&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16847059) JIRA comment. 

I always get "Could not find method call() for arguments [:runners:google-cloud-dataflow-java:worker] on project ':sdks:python:test-suites:dataflow:py35' of type org.gradle.api.Project". When the code remains split up (in three seperate build.gradle files), it does work. @tvalentyn any ideas?

**Comment by tvalentyn** [comment]
My gradle fu is not strong enough to answer what's missing without digging, @markflyhigh might be able to answer your question. Let's keep the code in separate files for now to unblock this, I can try to do the refactoring while working on https://issues.apache.org/jira/browse/BEAM-7063.

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

**Comment by fredo838** [comment]
Run Seed Job

**Comment by fredo838** [comment]
Run Python Dataflow ValidatesRunner

File: project-mappings (hunk 5 above)
Comment (tvalentyn) [reply project-mappings]: It is a temporary alias to help people transition from old ways to name gradle task to new names, see: https://github.com/apache/beam/blob/abece47cc1c1c88a519e54e67a2d358b439cf69c/deprecation-warning.txt

**Comment by tvalentyn** [comment]
Looking at Gradle scan https://scans.gradle.com/s/77k7ykxpwyiz4 on the latest run of ValidatesRunner tests on this PR, looks like Py3 suites did not run. Did we miss something?

**Comment by fredo838** [comment]
Run Seed Job

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
Tests failed due to a flake in Py35 suite due to GCE stockouts. Recent Gradle run included Py 3 tests,  but did not run Py 2 tests, instead the logs have a line: `"11:16:18 > Task :pythonValidatesRunner UP-TO-DATE
"` 
@markflyhigh  do you know whether gradle is caching the results of previous execution of integration test tasks and is not re-running them the second time?

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
Run Seed Job

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

File: build.gradle [hunk 6]
```diff
@@ -204,12 +204,25 @@ task python3PostCommit() {
...
   dependsOn ":sdks:python:test-suites:dataflow:py36:postCommitIT"
   dependsOn ":sdks:python:test-suites:dataflow:py37:postCommitIT"
-  dependsOn ":sdks:python:test-suites:dataflow:py35:validatesRunnerBatchTests"
   dependsOn ":sdks:python:test-suites:direct:py35:postCommitIT"
   dependsOn ":sdks:python:test-suites:direct:py36:postCommitIT"
   dependsOn ":sdks:python:test-suites:direct:py37:postCommitIT"
 }
 
+task pythonValidatesRunner() {
+  ":sdks:python:test-suites:dataflow:validatesRunnerBatchTests"
```
Comment (tvalentyn) [inline build.gradle]: Lines 213, 214 are missing dependsOn.

Comment (tvalentyn) [reply build.gradle]: I'll finish this in https://github.com/apache/beam/pull/8721.

**Pull request closed without merging**